src/Common/HashTable/HashTable.h

```diff
@@ -954,6 +958,67 @@ class HashTable :
         return const_cast<std::decay_t<decltype(*this)> *>(this)->find(x, hash_value);
     }
 
+    template <typename = std::enable_if<Grower::performs_linear_probing_with_single_step, void>>
+    void ALWAYS_INLINE erase(const Key & x)
+    {
+        /** Deletion from open addressing hash table without tombstones
+          *
+          * https://en.wikipedia.org/wiki/Linear_probing
+          * https://en.wikipedia.org/wiki/Open_addressing
+          * Algorithm without recomputing hash but keep probes difference value (difference of natural cell position and inserted one)
+          *  in cell https://arxiv.org/ftp/arxiv/papers/0909/0909.2547.pdf
+          *
+          * Currently we use algorithm with hash recomputing on each step from https://en.wikipedia.org/wiki/Open_addressing
+          */
+
+        if (Cell::isZero(x, *this))
+        {
+            if (this->hasZero())
+            {
+                --m_size;
+                this->clearHasZero();
+            }
+            else
+            {
+                return;
+            }
+        }
+
+        size_t hash_value = hash(x);
+        size_t i = findCell(x, hash_value, grower.place(hash_value));
+
+        if (buf[i].isZero(*this))
+        {
+            return;
+        }
+
+        /// We need to guarantee loop termination because there will be empty position
+        assert(m_size < grower.bufSize());
+
+        size_t j = i;
+
+        while (true)
+        {
+            j = grower.next(j);
+
+            if (buf[j].isZero(*this))
+            {
+                break;
+            }
+
+            size_t k = grower.place(buf[j].getHash(*this));
+
+            if (i <= j ? ((k <= i) || (k > j)) : ((k <= i) && (k > j)))
```

Comment:
memcpy implies `i != j` but here we check for `i <= j` that is confusing.

Reply:
It's strange to use ?: operator inside if condition.

&& and || will look better.

Reply:
Agreed.